File: python/src/trailproof/stores/memory.py
"""In-memory trail event store."""

from trailproof.chain import GENESIS_HASH
from trailproof.stores.base import TrailStore
from trailproof.types import QueryFilters, QueryResult, TrailEvent


class MemoryStore(TrailStore):
    """In-memory store that keeps events in a list.

    Events are lost when the process exits. Useful for testing
    and short-lived applications.
    """

    def __init__(self) -> None:
        self._events: list[TrailEvent] = []

    def append(self, event: TrailEvent) -> None:
        """Append a single event to the in-memory list."""
        self._events.append(event)

    def read_all(self) -> list[TrailEvent]:
        """Return all events in insertion order."""
        return list(self._events)

    def query(self, filters: QueryFilters) -> QueryResult:
        """Query events with optional filters and cursor pagination."""
        events = self._events

        # Apply cursor: skip events up to and including the cursor event_id
        if filters.cursor is not None:
            cursor_index = -1
            for i, event in enumerate(events):
                if event.event_id == filters.cursor:
                    cursor_index = i
                    break
            if cursor_index == -1:
                return QueryResult(events=[], next_cursor=None)
            events = events[cursor_index + 1 :]

        # Apply filters
        if filters.event_type is not None:
            events = [e for e in events if e.event_type == filters.event_type]
        if filters.actor_id is not None:
            events = [e for e in events if e.actor_id == filters.actor_id]
        if filters.tenant_id is not None:
            events = [e for e in events if e.tenant_id == filters.tenant_id]
        if filters.trace_id is not None:
            events = [e for e in events if e.trace_id == filters.trace_id]
        if filters.session_id is not None:
            events = [e for e in events if e.session_id == filters.session_id]
        if filters.from_time is not None:
            events = [e for e in events if e.timestamp >= filters.from_time]
        if filters.to_time is not None:
            events = [e for e in events if e.timestamp <= filters.to_time]

        # Apply limit and determine next_cursor
        limit = filters.limit
        if len(events) > limit:
            next_cursor = events[limit - 1].event_id
            events = events[:limit]
        else:
            next_cursor = None

        return QueryResult(events=events, next_cursor=next_cursor)

    def last_hash(self) -> str:
        """Return hash of the last event, or genesis hash if empty."""
        if not self._events:
            return GENESIS_HASH
        return self._events[-1].hash

    def count(self) -> int:
        """Return total number of stored events."""
        return len(self._events)

File: typescript/src/stores/memory.ts
/**
 * In-memory trail event store.
 */

import { GENESIS_HASH } from "../chain.js";
import type { QueryFilters, QueryResult, TrailEvent } from "../types.js";
import type { TrailStore } from "./base.js";

const DEFAULT_LIMIT = 100;

/**
 * In-memory store that keeps events in an array.
 *
 * Events are lost when the process exits. Useful for testing
 * and short-lived applications.
 */
export class MemoryStore implements TrailStore {
  private events: TrailEvent[] = [];

  /** Append a single event to the in-memory array. */
  append(event: TrailEvent): void {
    this.events.push(event);
  }

  /** Return all events in insertion order. */
  readAll(): TrailEvent[] {
    return [...this.events];
  }

  /** Query events with optional filters and cursor pagination. */
  query(filters: QueryFilters): QueryResult {
    let events: TrailEvent[] = this.events;

    // Apply cursor: skip events up to and including the cursor event_id
    if (filters.cursor != null) {
      const cursorIndex = events.findIndex((e) => e.event_id === filters.cursor);
      if (cursorIndex === -1) {
        return { events: [], next_cursor: null };
      }
      events = events.slice(cursorIndex + 1);
    }

    // Apply filters
    if (filters.event_type != null) {
      events = events.filter((e) => e.event_type === filters.event_type);
    }
    if (filters.actor_id != null) {
      events = events.filter((e) => e.actor_id === filters.actor_id);
    }
    if (filters.tenant_id != null) {
      events = events.filter((e) => e.tenant_id === filters.tenant_id);
    }
    if (filters.trace_id != null) {
      events = events.filter((e) => e.trace_id === filters.trace_id);
    }
    if (filters.session_id != null) {
      events = events.filter((e) => e.session_id === filters.session_id);
    }
    if (filters.from_time != null) {
      const fromTime = filters.from_time;
      events = events.filter((e) => e.timestamp >= fromTime);
    }
    if (filters.to_time != null) {
      const toTime = filters.to_time;
      events = events.filter((e) => e.timestamp <= toTime);
    }

    // Apply limit and determine next_cursor
    const limit = filters.limit ?? DEFAULT_LIMIT;
    if (events.length > limit) {
      const cursorEvent = events[limit - 1];
      const nextCursor = cursorEvent ? cursorEvent.event_id : null;
      events = events.slice(0, limit);
      return { events, next_cursor: nextCursor };
    }

    return { events, next_cursor: null };
  }

  /** Return hash of the last event, or genesis hash if empty. */
  lastHash(): string {
    if (this.events.length === 0) {
      return GENESIS_HASH;
    }
    const last = this.events[this.events.length - 1];
    return last ? last.hash : GENESIS_HASH;
  }

  /** Return total number of stored events. */
  count(): number {
    return this.events.length;
  }
}
